### PR TITLE
Streamline the TPF and lightcurve base classes for broad use cases

### PR DESCRIPTION
**Pull request by gully**
Will make it possible to use lightkurve on Spitzer data, for example.  E.g. #603

### CODE REVIEW

File: lightkurve/targetpixelfile.py [hunk 1]
```diff
@@ -1464,6 +1505,10 @@ def _make_aperture_extension(self):
         hdu.header['EXTNAME'] = 'APERTURE'
         return hdu
 
+class KeplerTargetPixelFileFactory(TargetPixelFileFactory):
+    """Same as TargetPixelFileFactory."""
+    # Backwards compatibility
+    pass
```
Comment (barentsen) [inline lightkurve/targetpixelfile.py]: 👍 Let's raise a depreciation warning in the constructor here!

File: lightkurve/targetpixelfile.py [hunk 2]
```diff
@@ -374,6 +384,48 @@ def to_lightcurve(self, method='aperture', **kwargs):
...
+            that the pixel will be used.
+            If None or 'all' are passed, all pixels will be used.
+            If 'pipeline' is passed, the mask suggested by the official pipeline
+            will be returned.
+            If 'threshold' is passed, all pixels brighter than 3-sigma above
+            the median flux will be used.
+
+        Returns
+        -------
+        lc : KeplerLightCurve object
```
Comment (barentsen) [inline lightkurve/targetpixelfile.py]: Suggested change to fix the return type and render nicely:
```suggestion
        lc : `.LightCurve` object
```

File: lightkurve/targetpixelfile.py [hunk 3]
```diff
@@ -374,6 +384,48 @@ def to_lightcurve(self, method='aperture', **kwargs):
...
+
+        keys = {'centroid_col': centroid_col,
+                'centroid_row': centroid_row,
+                'quality': self.quality,
+                'channel': self.channel,
+                'ra': self.ra,
+                'dec': self.dec,
+                'label': self.header['OBJECT'],
+                'targetid': self.targetid}
+        return LightCurve(time=self.time,
```
[the remark this answers is not in the document]
Comment (barentsen) [reply lightkurve/targetpixelfile.py]: I agree with @codacy-bot here: a number of arguments are being passed to LightCurve via `**keys` that are not valid keyword arguments (ra, dec, channel, quality, centroid_col, and centroid_row).

File: lightkurve/targetpixelfile.py [hunk 4]
```diff
@@ -944,33 +991,19 @@ def extract_aperture_photometry(self, aperture_mask='pipeline'):
...
-                'ra': self.ra,
-                'dec': self.dec,
-                'label': self.header['OBJECT'],
-                'targetid': self.targetid}
-        return KeplerLightCurve(time=self.time,
-                                flux=np.nansum(self.flux[:, aperture_mask], axis=1),
-                                flux_err=flux_err,
+                'mission': self.mission,
+                'channel': self.channel}
+        keys.update(kepler_keys)
```
Comment (barentsen) [inline lightkurve/targetpixelfile.py]: I don't understand why we are using a `keys` dictionary here.  I'd get rid of `keys` altogether and, instead, simply pass the arguments directly via the constructor of `KeplerLightCurve` being called on the line below.

**Comment by gully** [comment]
cadenceno, centroids, ra, dec, and quality can all go into the base class of lightcurve.  Things that cannot go into the base class are truly mission specific: "sector", "camera", "quarter", "campaign", "module", "channel", "ccd", etc.

**Comment by gully** [comment]
Update: cadenceno should *not* be in the generic LightCurve class because its carries special meaning-- it confers presence of gaps in data, and how disconnected observations connect.

**Comment by gully** [comment]
Woohoo, interact now works with the most recent commits.  One hiccup is how to support a wide range to time precisions.  I think we should detect these in the factory and put them into the custom TPF headers.  We want to avoid `if self.mission=='blah'` since the base TPF should be more-or-less agnostic to missions.

**Comment by gully** [comment]
Tests are failing due to a python2 hiccup with `super()`.  Will investigate later, along with Geert's excellent feedback thank you Geert!

**Comment by gully** [comment]
This PR is getting close.  Still todo:
- [ ] Fix a bunch of `TODO` breadcrumbs left around
- [ ] Add a real docstring to the TPF baseclass
- [ ] Add some tests for TPF baseclass

**Comment by gully** [comment]
New goal is to move all the factory into its own module.

File: lightkurve/factory.py [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+            mission = hdu0_keywords['MISSION']
+
+        if (mission=='Kepler') or (mission=='K2'):
+            return KeplerTargetPixelFile(self._hdulist(hdu0_keywords=hdu0_keywords,
+                                        ext_info=ext_info), **kwargs)
+        elif mission=='TESS':
+            return TessTargetPixelFile(self._hdulist(hdu0_keywords=hdu0_keywords,
+                                        ext_info=ext_info), **kwargs)
+        else:
+            return TargetPixelFile(self._hdulist(hdu0_keywords=hdu0_keywords,
```
Comment (barentsen) [inline lightkurve/factory.py]: Codacy Issue found: [undefined name 'TargetPixelFile'](https://app.codacy.com/app/barentsen/lightkurve/pullRequest?prid=4287604)

File: lightkurve/targetpixelfile.py [hunk 6]
```diff
@@ -23,31 +23,45 @@
...
-from .lightcurve import KeplerLightCurve, TessLightCurve
+from .lightcurve import KeplerLightCurve, TessLightCurve, LightCurve
 from .prf import KeplerPRF
 from .utils import KeplerQualityFlags, TessQualityFlags, \
                    plot_image, bkjd_to_astropy_time, btjd_to_astropy_time, \
                    LightkurveWarning, detect_filetype, validate_method, \
                    centroid_quadratic
 
-__all__ = ['KeplerTargetPixelFile', 'TessTargetPixelFile']
+from .factory import KeplerTargetPixelFileFactory, TargetPixelFileFactory
```
Comment (barentsen) [inline lightkurve/targetpixelfile.py]: Codacy Issue found: ['.factory.KeplerTargetPixelFileFactory' imported but unused](https://app.codacy.com/app/barentsen/lightkurve/pullRequest?prid=4287604)

File: lightkurve/factory.py [hunk 7]
```diff
@@ -0,0 +1,285 @@
...
+
+        mission = None
+        if 'MISSION' in hdu0_keywords.keys():
+            mission = hdu0_keywords['MISSION']
+
+        if (mission=='Kepler') or (mission=='K2'):
+            return KeplerTargetPixelFile(self._hdulist(hdu0_keywords=hdu0_keywords,
+                                        ext_info=ext_info), **kwargs)
+        elif mission=='TESS':
+            return TessTargetPixelFile(self._hdulist(hdu0_keywords=hdu0_keywords,
```
Comment (barentsen) [inline lightkurve/factory.py]: Codacy Issue found: [undefined name 'TessTargetPixelFile'](https://app.codacy.com/app/barentsen/lightkurve/pullRequest?prid=4287604)

File: lightkurve/factory.py [hunk 8]
```diff
@@ -0,0 +1,285 @@
...
+from __future__ import division
+import datetime
+import os
+import warnings
+import logging
+
+from astropy.io import fits
+
+import numpy as np
+from tqdm import tqdm
```
Comment (barentsen) [inline lightkurve/factory.py]: Codacy Issue found: [Unused tqdm imported from tqdm](https://app.codacy.com/app/barentsen/lightkurve/pullRequest?prid=4287604)

File: lightkurve/factory.py [hunk 9]
```diff
@@ -0,0 +1,285 @@
...
+            return TargetPixelFile(self._hdulist(hdu0_keywords=hdu0_keywords,
+                                        ext_info=ext_info), **kwargs)
+
+    def _hdulist(self, hdu0_keywords, ext_info):
+        """Returns an astropy.io.fits.HDUList object."""
+        return fits.HDUList([self._make_primary_hdu(hdu0_keywords=hdu0_keywords),
+                             self._make_target_extension(ext_info=ext_info),
+                             self._make_aperture_extension()])
+
+    def _header_template(self, extension):
```
Comment (barentsen) [inline lightkurve/factory.py]: Codacy Issue found: [Method could be a function](https://app.codacy.com/app/barentsen/lightkurve/pullRequest?prid=4287604)

File: lightkurve/tests/test_targetpixelfile.py [hunk 10]
```diff
@@ -207,14 +214,20 @@ def test_repr():
...
                 TessTargetPixelFile(filename_tess)]:
         tpf.to_lightcurve()
         tpf.to_lightcurve(aperture_mask=None)
         tpf.to_lightcurve(aperture_mask='all')
         lc = tpf.to_lightcurve(aperture_mask='pipeline')
-        assert lc.astropy_time.scale == 'tdb'
         assert lc.label == tpf.hdu[0].header['OBJECT']
+        # Generic TPFs have unknown timescales by default
+        if tpf.__class__ == TargetPixelFile:
+            assert lc.time_scale == None
```
Comment (barentsen) [inline lightkurve/tests/test_targetpixelfile.py]: Codacy Issue found: [comparison to None should be 'if cond is None:'](https://app.codacy.com/app/barentsen/lightkurve/pullRequest?prid=4287604)

File: lightkurve/targetpixelfile.py (hunk 6 above)
Comment (barentsen) [inline lightkurve/targetpixelfile.py]: Codacy Issue found: [Unused KeplerTargetPixelFileFactory imported from factory](https://app.codacy.com/app/barentsen/lightkurve/pullRequest?prid=4287604)

File: lightkurve/tests/test_targetpixelfile.py [hunk 11]
```diff
@@ -207,14 +214,20 @@ def test_repr():
...
+            assert lc.time_format == None
```
Comment (barentsen) [inline lightkurve/tests/test_targetpixelfile.py]: Codacy Issue found: [comparison to None should be 'if cond is None:'](https://app.codacy.com/app/barentsen/lightkurve/pullRequest?prid=4287604)

File: lightkurve/factory.py [hunk 12]
```diff
@@ -0,0 +1,285 @@
...
+        if ext_info is None:
+            ext_info = {}
+        self._check_data()
+
+        mission = None
+        if 'MISSION' in hdu0_keywords.keys():
+            mission = hdu0_keywords['MISSION']
+
+        if (mission=='Kepler') or (mission=='K2'):
+            return KeplerTargetPixelFile(self._hdulist(hdu0_keywords=hdu0_keywords,
```
Comment (barentsen) [inline lightkurve/factory.py]: Codacy Issue found: [undefined name 'KeplerTargetPixelFile'](https://app.codacy.com/app/barentsen/lightkurve/pullRequest?prid=4287604)

File: lightkurve/factory.py (hunk 8 above)
Comment (barentsen) [inline lightkurve/factory.py]: Codacy Issue found: ['tqdm.tqdm' imported but unused](https://app.codacy.com/app/barentsen/lightkurve/pullRequest?prid=4287604)

File: lightkurve/factory.py (hunk 9 above)
Comment (gully) [reply lightkurve/factory.py]: Meh?

File: lightkurve/factory.py [hunk 13]
```diff
@@ -0,0 +1,285 @@
...
+            return TargetPixelFile(self._hdulist(hdu0_keywords=hdu0_keywords,
+                                        ext_info=ext_info), **kwargs)
+
+    def _hdulist(self, hdu0_keywords, ext_info):
+        """Returns an astropy.io.fits.HDUList object."""
+        return fits.HDUList([self._make_primary_hdu(hdu0_keywords=hdu0_keywords),
+                             self._make_target_extension(ext_info=ext_info),
+                             self._make_aperture_extension()])
+
+    def _header_template(self, extension):
```
Comment (barentsen) [inline lightkurve/factory.py]: Codacy Issue found: [Method could be a function](https://app.codacy.com/app/barentsen/lightkurve/pullRequest?prid=4287604)

File: lightkurve/tests/test_targetpixelfile.py [hunk 14]
```diff
@@ -18,8 +18,9 @@
 import astropy.units as u
 from astropy.utils.exceptions import AstropyWarning
 
-from ..targetpixelfile import KeplerTargetPixelFile, KeplerTargetPixelFileFactory
-from ..targetpixelfile import TessTargetPixelFile
+from ..targetpixelfile import KeplerTargetPixelFile, TargetPixelFileFactory
+from ..targetpixelfile import TessTargetPixelFile, TargetPixelFile
+from ..factory import TargetPixelFileFactory
```
Comment (barentsen) [inline lightkurve/tests/test_targetpixelfile.py]: Codacy Issue found: [redefinition of unused 'TargetPixelFileFactory' from line 21](https://app.codacy.com/app/barentsen/lightkurve/pullRequest?prid=4287604)

File: lightkurve/factory.py [hunk 15]
```diff
@@ -0,0 +1,286 @@
...
```
Comment (barentsen) [inline lightkurve/factory.py]: Codacy Issue found: [Method could be a function](https://app.codacy.com/app/barentsen/lightkurve/pullRequest?prid=4287604)

File: lightkurve/tests/test_targetpixelfile.py (hunk 14 above)
Comment (barentsen) [inline lightkurve/tests/test_targetpixelfile.py]: Codacy Issue found: [Reimport 'TargetPixelFileFactory' (imported line 21)](https://app.codacy.com/app/barentsen/lightkurve/pullRequest?prid=4287604)

**Comment by gully** [comment]
Woohoo, checks have passed.
I had originally made changes to add `time_format` to TPF classes, but have since decided not to do that, since there's no generic way to determine the time format from arbitrary data.  It's possible we could add it as a kwarg to init the TPF, but let's leave that for later if we need it at all.

File: lightkurve/interact.py [hunk 16]
```diff
@@ -53,29 +53,37 @@ def prepare_lightcurve_datasource(lc):
...
     lc_source : bokeh.plotting.ColumnDataSource
     """
-    # Convert time into human readable strings, breaks with NaN time
-    # See https://github.com/KeplerGO/lightkurve/issues/116
-    if (lc.time == lc.time).all():
+    # Convert time into human readable strings
+    try:
         human_time = lc.astropy_time.isot
-    else:
+    except:
```
Comment (barentsen) [inline lightkurve/interact.py]: Codacy Issue found: [No exception type(s) specified](https://app.codacy.com/app/barentsen/lightkurve/pullRequest?prid=4287604)

File: lightkurve/targetpixelfile.py [hunk 17]
```diff
@@ -14,6 +13,7 @@
 from astropy.utils.exceptions import AstropyWarning
 from astropy.coordinates import SkyCoord
 from astropy.stats.funcs import median_absolute_deviation as MAD
+from astropy.time import Time
```
Comment (barentsen) [inline lightkurve/targetpixelfile.py]: Codacy Issue found: [Unused Time imported from astropy.time](https://app.codacy.com/app/barentsen/lightkurve/pullRequest?prid=4287604)

Comment (barentsen) [inline lightkurve/targetpixelfile.py]: Codacy Issue found: ['astropy.time.Time' imported but unused](https://app.codacy.com/app/barentsen/lightkurve/pullRequest?prid=4287604)

**Comment by gully** [comment]
Rebased to master again.

File: lightkurve/targetpixelfile.py [hunk 18]
```diff
@@ -867,6 +937,151 @@ def cutout(self, center=None, size=5):
             newfits = fits.HDUList(hdus)
         return self.__class__(newfits)
 
+    def from_fits_images(images, position, size=(11, 11), extension=1,
```
Comment (barentsen) [inline lightkurve/targetpixelfile.py]: Codacy Issue found: [TargetPixelFile.from_fits_images is too complex (19)](https://app.codacy.com/app/barentsen/lightkurve/pullRequest?prid=4287604)

File: lightkurve/targetpixelfile.py [hunk 19]
```diff
@@ -867,6 +926,151 @@ def cutout(self, center=None, size=5):
...
```
Comment (barentsen) [inline lightkurve/targetpixelfile.py]: Codacy Issue found: [TargetPixelFile.from_fits_images is too complex (19)](https://app.codacy.com/app/barentsen/lightkurve/pullRequest?prid=4287604)

**Comment by gully** [comment]
Thank you @christinahedges for the review.  I have made all the requested changes in the review, except for one, which is to move the static method `from_fits_images` out of the `TargetPixelFile` class and into the new factory module, presumably as a (*non-*static) method of `TargetPixelFileFactory`.  The relevant code block begins with:

```
@staticmethod
def from_fits_images(images, position, size=(11, 11), extension=1, [...])
```

I agree that this proposed code change is the right way to go-- lots of the `from_fits_images` functionality hinges on the factory, and it contains significant mission-specific keywords that make it unsuitable for a generic TPF base class.  The lightkurve design is moving away from static methods.  Eventually we want to overhaul the factory to include mission-specific dataloaders, so the already-oversized `from_fits_images` could be split into bite-size sub-functions that are easier to maintain and test, etc.

I started down the path of migrating `from_fits_images` to `factory.py`, and encountered some friction that made it more complicated than a mere copy-and-paste.  Here are the chief problems I encountered:

1. Changes will be masked by existing refactor
Christina noted that this PR splits out the relatively large `TargetPixelFileFactory` code block into a new file, masking any changes from its heritage, and making it difficult to review the changes without a granular inspect of each commit or groups of commits, perhaps with intermediate rebasing.  Currently the squashed commit will be relatively free of changes to the factory, but in the proposed `from_fits_images` relocation, the content differences will be significant and hard to follow.

2. Breaking backwards compatibility
The current PR preserves backwards compatibility--existing tutorials and code that use `from_fits_images` will still work everywhere, making the testing requirements for this refactor PR relatively low.  The proposed `from_fits_images` change would break any existing code that used the factory or from_fits_images.  These functions are esoteric/advanced enough to safely assume that few users would notice, but we'd still want to change the tutorials, add a deprecation warning, add some testing, and revise existing tests.

3. Even a small redesign motivates an overhaul of the factory
The `from_fits_images` arguments are at odds with the existing `TargetPixelFileFactory` init signature: arguments are duplicated, the passed `**kwargs` takes on new meaning, and other subtle differences.  `from_fits_images` can be significantly streamlined by taking advantage of being a method on TargetPixelFileFactory.  In contemplating these compatibility changes, it may be simpler to start with a fresh redesign of the Factory and the data loader helper functions.  

For these reasons I propose that the existing PR leave the sub-optimal `from_fits_images` in place, and a new PR can begin the more-involved process of migrating it to the factory, and overhauling the data loading process.

**Comment by gully** [comment]
Hey all, I'm going to be building off this this PR.  I know @barentsen was reluctant to merge since the PR touches several files.  Rather than merge or delay the work, I will follow the following workflow:
- Submit a PR to my fork, basing off this branch.
- Once the current PR is merged, I can change the base from `gully/spitzer_tpfs` to `lightkurve/master`

File: lightkurve/interact.py [hunk 20]
```diff
@@ -215,6 +228,8 @@ def add_gaia_figure_elements(tpf, fig, magnitude_limit=18):
     pix_scale = 4.0  # arcseconds / pixel for Kepler, default
     if tpf.mission == 'TESS':
         pix_scale = 21.0
+    if tpf.mission == 'Spitzer':
+        pix_scale = 1.22
```
Comment (barentsen) [inline lightkurve/interact.py]: Can we get the pixel scale from the TPF FITS header, e.g. one of the standard WCS keywords?  I don't super-like hard-coding it here.

File: lightkurve/interact.py [hunk 21]
```diff
@@ -53,29 +53,37 @@ def prepare_lightcurve_datasource(lc):
...
+            flag_str_list = KeplerQualityFlags.decode(bitmask)
+            if len(flag_str_list) == 0:
+                qual_strings.append(' ')
+            if len(flag_str_list) == 1:
+                qual_strings.append(flag_str_list[0])
+            if len(flag_str_list) > 1:
+                qual_strings.append("; ".join(flag_str_list))
+    else:
+        qual_strings = [' '] * len(lc.flux)
+        #TODO add TESS support
```
Comment (barentsen) [inline lightkurve/interact.py]: Let's support TESS here properly while at it; it should only involve using `TessQualityFlags` instead of `KeplerQualityFlags` above.

File: lightkurve/lightcurve.py [hunk 22]
```diff
@@ -72,7 +84,9 @@ class LightCurve(object):
     """
     def __init__(self, time=None, flux=None, flux_err=None, flux_unit=None,
                  time_format=None, time_scale=None, targetid=None, label=None,
-                 meta=None):
+                 centroid_col=None, centroid_row=None, quality=None,
+                 quality_bitmask=None, mission=None,ra=None,
+                 dec=None, meta=None):
```
Comment (barentsen) [inline lightkurve/lightcurve.py]: Note: `ra` and `dec` are not listed in the docstring above.

File: lightkurve/lightcurve.py [hunk 23]
```diff
@@ -1703,26 +1729,19 @@ def __init__(self, time=None, flux=None, flux_err=None,
                  cadenceno=None, targetid=None, ra=None, dec=None, label=None, meta=None):
         super(KeplerLightCurve, self).__init__(time=time, flux=flux, flux_err=flux_err, flux_unit=flux_unit,
                                                time_format=time_format, time_scale=time_scale,
-                                               targetid=targetid, label=label, meta=meta)
-        self.centroid_col = self._validate_array(centroid_col, name='centroid_col')
-        self.centroid_row = self._validate_array(centroid_row, name='centroid_row')
-        self.quality = self._validate_array(quality, name='quality')
+                                               targetid=targetid, label=label, centroid_col=centroid_col,
+                                               centroid_row=centroid_row, quality=quality,
+                                               quality_bitmask=quality_bitmask,ra=ra, dec=dec,meta=meta)
```
Comment (barentsen) [inline lightkurve/lightcurve.py]:
```suggestion
                                               quality_bitmask=quality_bitmask, ra=ra, dec=dec,meta=meta)
```

File: lightkurve/lightcurve.py [hunk 24]
```diff
@@ -1854,27 +1873,20 @@ def __init__(self, time=None, flux=None, flux_err=None,
...
         super(TessLightCurve, self).__init__(time=time, flux=flux, flux_err=flux_err, flux_unit=flux_unit,
-                                             time_format=time_format, time_scale=time_scale,
-                                             targetid=targetid, label=label, meta=meta)
-        self.centroid_col = self._validate_array(centroid_col, name='centroid_col')
-        self.centroid_row = self._validate_array(centroid_row, name='centroid_row')
-        self.quality = self._validate_array(quality, name='quality')
+                                               time_format=time_format, time_scale=time_scale,
+                                               targetid=targetid, label=label, centroid_col=centroid_col,
+                                               centroid_row=centroid_row, quality=quality,
+                                               quality_bitmask=quality_bitmask,ra=ra, dec=dec,meta=meta)
```
Comment (barentsen) [inline lightkurve/lightcurve.py]:
```suggestion
                                               quality_bitmask=quality_bitmask, ra=ra, dec=dec,meta=meta)
```

File: lightkurve/targetpixelfile.py [hunk 25]
```diff
@@ -375,6 +399,42 @@ def to_lightcurve(self, method='aperture', **kwargs):
         else:
             raise ValueError("Photometry method must be 'aperture' or 'prf'.")
 
+    def extract_aperture_photometry(self, aperture_mask='pipeline', centroid_method='moments'):
```
Comment (barentsen) [inline lightkurve/targetpixelfile.py]: Just as a heads up: I think I moved `extract_aperture_photometry` to the base class in #651 last week to resolve a bug, so this may trigger a complicated merge conflict.  Apologies!

**Comment by barentsen** [comment]
I have consolidated this and the related PRs into a single, new PR, see here:
https://github.com/KeplerGO/lightkurve/pull/678
